tests/specs/controllers/EAppReferenceController.test.js

```diff
@@ -0,0 +1,92 @@
+const { expect } = require('chai');
+const sinon = require('sinon');
+
+const EAppReferenceController = require('../../../api/controllers/EAppReferenceController');
+
+describe('EAppReferenceController', () => {
+    let reqStub = {};
+    let resStub = {};
+    const sandbox = sinon.sandbox.create();
+
+    beforeEach(() => {
+        reqStub = {
+            body: {
+                'user-reference': 136542,
+            },
+            session: {
+                eApp: {
+                    userRef: 136542,
+                },
+            },
+        };
+        resStub = {
+            serverError: sandbox.spy(),
+            response: sandbox.spy(),
+            view: sandbox.spy(),
+            redirect: sandbox.spy(),
+        };
+
+        sandbox.spy(sails.log, 'info');
+        sandbox.spy(sails.log, 'error');
+    });
+
+    afterEach(() => {
+        sandbox.restore();
+    });
+
+    describe('renderPage', () => {
+        it('prevents users from seeing the page if they are not logged in', () => {
+            // when
+            sandbox.stub(HelperService, 'getUserData').callsFake(() => ({
+                loggedIn: false,
+            }));
+            EAppReferenceController.renderPage(reqStub, resStub);
+
+            // then
+            expect(
+                sails.log.error.getCall(0).args[0] ===
+                    'User not logged in'
```

Comment:
Could you think of a way to test without relying on spying on the log? When implementation details are tested, the tests end up being flakier (likely to break down the line), and that will affect time it takes for a feature to be "done" (since you have to go back and fix the tests). 

it looks like spying on sails.log is a workaround for testing this "error" block in `EAppReferenceController.renderPage`.  Ideally there shouldn't be a need to test a 3rd party library (i.e. sails.log) because it's already well tested. You may need to spy and test on sails.log once in your application to make sure it can be called.

from Kent C Dodds ((react) testing-library author) 
>Implementation details are things which users of your code will not typically use, see, or even know about.

Here is his blog post about it (however this one is more about front end testing https://kentcdodds.com/blog/testing-implementation-details) 

You sort of have to imagine you only know the parameters of the method (req, res), does the method return what you want it to? and if renderPage requires too much stubbing/spying, how could you test smaller chunks of the code?

If a method doesn't return anything, yes you may have to resort to spying on a log

Reply:
That makes sense, I think for me it's always easier to figure out how to write FE tests because they're more visual but BE tests I think I need to get the hang of.

Also, I know you mentioned this in the previous PR, hopefully, I'll have this down in the next PR I create :)